File: src/main/java/com/gmail/borlandlp/minigamesdtools/nmsentities/PathController.java
package com.gmail.borlandlp.minigamesdtools.nmsentities;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.nmsentities.classes.PathProvider;
import com.gmail.borlandlp.minigamesdtools.nmsentities.entity.IMoveControllableEntity;

public class PathController {
    private PathProvider[] paths;
    private int curIndex = 0;
    private IMoveControllableEntity entity;
    private boolean isFinished;
    private boolean isRepeating;
    int failCounter;

    public PathController(IMoveControllableEntity entity, PathProvider[] paths, boolean isRepeating) {
        this.paths = paths;
        this.entity = entity;
        this.isRepeating = isRepeating;

        this.getEntity().setPath(this.getCurrentPath());
    }

    public PathProvider[] getPaths() {
        return this.paths;
    }

    public boolean isFinished() {
        return isFinished;
    }

    public IMoveControllableEntity getEntity() {
        return entity;
    }

    public boolean isRepeating() {
        return isRepeating;
    }

    public PathProvider getCurrentPath() {
        return this.getPaths()[this.curIndex];
    }

    public void update() throws Exception {
        if(this.isFinished()) {
            return;
        } else if(this.curIndex >= this.getPaths().length) {
            throw new Exception("index out of bounds!");
        }

        if(this.getPaths()[this.curIndex].pathEnding()) {//if the entity has reached the path point
            // next path
            if((this.curIndex + 1) >= this.getPaths().length) {
                if(this.isRepeating()) {
                    this.curIndex = 0;
                } else {
                    this.isFinished = true;
                    return;
                }
            } else {
                this.curIndex++;
            }

            PathProvider nextPath = this.getCurrentPath();
            if(this.isRepeating() && nextPath.pathEnding()) {
                nextPath.reloadPath();
            }
            this.failCounter = 0;
            this.getEntity().setPath(nextPath);

        } else if(!this.getEntity().isFollowsThisPath(this.getCurrentPath())) { // if entity switched path
            if(++this.failCounter > 5) {
                Debug.print(Debug.LEVEL.NOTICE, "teleport entity");
                this.failCounter = 0;
                this.getEntity().teleport(this.getCurrentPath());
                this.getEntity().setPath(this.getCurrentPath());
            }
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/nmsentities/PathController.kt
package com.gmail.borlandlp.minigamesdtools.nmsentities

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.nmsentities.classes.PathProvider
import com.gmail.borlandlp.minigamesdtools.nmsentities.entity.IMoveControllableEntity

class PathController(
    val entity: IMoveControllableEntity,
    val paths: Array<PathProvider>,
    val isRepeating: Boolean
) {
    private var curIndex = 0
    var isFinished = false
        private set
    var failCounter = 0

    val currentPath: PathProvider
        get() = paths[curIndex]

    @Throws(Exception::class)
    fun update() {
        if (isFinished) {
            return
        } else if (curIndex >= paths.size) {
            throw Exception("index out of bounds!")
        }
        if (paths[curIndex].pathEnding()) { //if the entity has reached the path point
// next path
            if (curIndex + 1 >= paths.size) {
                if (isRepeating) {
                    curIndex = 0
                } else {
                    isFinished = true
                    return
                }
            } else {
                curIndex++
            }
            val nextPath = currentPath
            if (isRepeating && nextPath.pathEnding()) {
                nextPath.reloadPath()
            }
            failCounter = 0
            entity.setPath(nextPath)
        } else if (!entity.isFollowsThisPath(currentPath)) { // if entity switched path
            if (++failCounter > 5) {
                Debug.print(
                    Debug.LEVEL.NOTICE,
                    "teleport entity"
                )
                failCounter = 0
                entity.teleport(currentPath)
                entity.setPath(currentPath)
            }
        }
    }

    init {
        entity.setPath(currentPath)
    }
}